### PR TITLE
Implement decoding API for specimen description

### PR DESCRIPTION
**Pull request by hackermd**
The PR provides a decoding API for the following classes:

- `highdicom.SpecimenDescription`
- `highdicom.SpecimenPreparationStep`
- `highdicom.SpecimenCollection`
- `highdicom.SpecimenProcesing`
- `highdicom.SpecimenSampling`
- `highdicom.SpecimenStaining`

Note that a breaking change is introduced for `highdicom.SpecimenPreparationStep` (see section in the release notes for details).

### CODE REVIEW

File: src/highdicom/content.py [hunk 1]
```diff
@@ -676,6 +717,12 @@ def __init__(
                 )
             self.append(item)
 
+    @property
+    def substances(self) -> List[CodedConcept]:
```
Comment (CPBridge) [inline src/highdicom/content.py:721]: This is not related to your changes here but I noticed that there's no check in the constructor that the passed list of substances is not empty,  which although I haven't checked the standard is a mistake I assume and should probably be fixed.

File: src/highdicom/content.py [hunk 2]
```diff
@@ -805,21 +869,216 @@ def __init__(
...
+
+        """
+        instance = deepcopy(dataset)
+        sequence = ContentSequence.from_sequence(
+            dataset.SpecimenPreparationStepContentItemSequence,
+            is_root=False,
+            is_sr=False
+        )
+        instance.SpecimenPreparationStepContentItemSequence = sequence
+        instance.__class__ = SpecimenPreparationStep
```
Comment (CPBridge) [inline src/highdicom/content.py:973]:
```suggestion
        instance.__class__ = cls
```

**Review by CPBridge — CHANGES_REQUESTED** [review]
A few little things here and there, mostly related to the changes to ImageFileReader.

The specimen decoding API looks nice. However I am confused about why rebasing `SpecimenPreparationStep` on `Dataset` (a breaking change, albeit one that is unlikely to affect many people) is necessary. There are several SR templates based on `ContentSequence` for which we have successfully built decoding APIs. Might be missing something but I can't see why this would be any different. Can you clarify?

File: src/highdicom/io.py [hunk 3]
```diff
@@ -233,16 +231,49 @@ class ImageFileReader(object):
 
     """
 
-    def __init__(self, filename: str):
+    def __init__(self, fp: Union[str, Path, DicomFileLike]):
```
[the remark this answers is not in the document]
Comment (hackermd) [reply src/highdicom/io.py:234]: Good catch! We can probably keep `filename`. It's not ideal but better than breaking the API.

File: src/highdicom/io.py [hunk 4]
```diff
@@ -280,26 +311,76 @@ def open(self) -> None:
...
+        Raises
+        ------
+        InvalidDicomError
+            If the file object does not represent a DICOM Part 10 file
+
+        """
+        # pos = fp.tell()
+        # read_preamble(fp, False)
+        # file_meta = _read_file_meta_info(fp)
+        # fp.seek(pos)
```
[the remark this answers is not in the document]
Comment (hackermd) [reply src/highdicom/io.py:353]: Addressed via ce352150b8602170d01a4047db6fab8b984ab68e

File: src/highdicom/io.py [hunk 5]
```diff
@@ -233,16 +231,49 @@ class ImageFileReader(object):
...
+                )
+            self._fp = fp
+            self._filename = None
+        elif isinstance(fp, (str, Path)):
+            self._filename = Path(fp)
+            self._fp = None
+        else:
+            raise TypeError(
+                'Argument "filename" must either an open DICOM file object or '
+                'the path to a DICOM file stored on disk.'
```
[the remark this answers is not in the document]
Comment (hackermd) [reply src/highdicom/io.py:270]: Addressed via ce352150b8602170d01a4047db6fab8b984ab68e

File: src/highdicom/io.py [hunk 6]
```diff
@@ -233,16 +231,49 @@ class ImageFileReader(object):
...
         Parameters
         ----------
-        filename: str
-            Path to a DICOM Part10 file containing a dataset of an image
-            SOP Instance
+        fp: Union[str, pathlib.Path, pydicom.filebase.DicomfileLike]
+            DICOM Part10 file containing a dataset of an image SOP Instance
 
         """
-        self.filename = filename
```
[the remark this answers is not in the document]
Comment (hackermd) [reply src/highdicom/io.py:245]: The filename can be determined from the file pointer via the `name` attribute. Addressed via ce352150b8602170d01a4047db6fab8b984ab68e

File: src/highdicom/io.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (hackermd) [reply src/highdicom/io.py]: Addressed via c517ed7bb7809deac72231f670381a6746b5aaa4

File: src/highdicom/io.py [hunk 7]
```diff
@@ -65,25 +66,24 @@ def _get_bot(fp: DicomFileLike, number_of_frames: int) -> List[int]:
     first_frame_offset = fp.tell()
     tag = TupleTag(fp.read_tag())
     if int(tag) != ItemTag:
-        raise ValueError('Reading of Basic Offset Table failed')
+        raise ValueError(
+            # 'Reading of Basic Offset Table failed. '
+            f'Encountered unexpected Tag "{tag}".'
+        )
```
[the remark this answers is not in the document]
Comment (hackermd) [reply src/highdicom/io.py]: Ups...fixed via af4c76e4a8c64a5c6ddfd15fa1123d36990fe9f0

File: src/highdicom/content.py (hunk 1 above)
Comment (hackermd) [reply src/highdicom/content.py:725]: You are right, there should be at least one substance. Addressed via cfa78c1779cc49e9e60ec382a220bd630714fdb7

File: src/highdicom/io.py [hunk 8]
```diff
@@ -280,26 +311,76 @@ def open(self) -> None:
...
+        ------
+        InvalidDicomError
+            If the file object does not represent a DICOM Part 10 file
+
+        """
+        # pos = fp.tell()
+        # read_preamble(fp, False)
+        # file_meta = _read_file_meta_info(fp)
+        # fp.seek(pos)
+        file_meta = read_file_meta_info(str(self._filename))
```
[the remark this answers is not in the document]
Comment (hackermd) [reply src/highdicom/io.py:350]: I think this has to be the path to the file:

https://github.com/pydicom/pydicom/blob/master/pydicom/filereader.py#L659-L668

Unfortunately, there is a `read_file_meta_info()` and a `_read_file_meta_info()` function. The former takes a string as input and the second a file pointer object.

File: src/highdicom/sr/value_types.py [hunk 9]
```diff
@@ -612,7 +646,7 @@ def _check_dataset(
                 f'Item #{index} of sequence is not an SR Content Item:\n'
                 f'{dataset}'
             )
-        if not hasattr(dataset, 'RelationshipType') and not is_root:
+        if not hasattr(dataset, 'RelationshipType') and not is_root and is_sr:
```
[the remark this answers is not in the document]
Comment (hackermd) [reply src/highdicom/sr/value_types.py:649]: Addressed via 26a3ef2014f61dfb41b86f3501e73c936d5aae92

File: src/highdicom/io.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (CPBridge) [reply src/highdicom/io.py:350]: I'll trust you have it figured out

File: src/highdicom/content.py [hunk 10]
```diff
@@ -566,16 +567,26 @@ def __init__(
         Parameters
         ----------
         procedure: Union[pydicom.sr.coding.Code, highdicom.sr.CodedConcept]
-            Procedure used to collect the examined specimen
+            Surgical procedure used to collect the examined specimen
 
         """  # noqa: E501
-        super().__init__(is_root=True)
+        super().__init__(is_root=False, is_sr=False)
```
[the remark this answers is not in the document]
Comment (hackermd) [reply src/highdicom/content.py:573]: Addressed via 06f7bc9a0010423fde4304e570227a2715d70a40

**Comment by CPBridge** [comment]
All looks good (except the style error), but I'm still curious to know, if only for my own sake, why the rebase of the `SpecimenPreparationStep` is necessary

**Review by CPBridge — APPROVED** [review]
Sorry, missed that you had fixed this

**Pull request merged**